python/ray/data/impl/table_block.py

```diff
@@ -0,0 +1,231 @@
+
+    def _concat_tables(self, tables: List[Block]) -> Block:
+        raise NotImplementedError
+
+    @staticmethod
+    def _empty_table() -> Any:
+        raise NotImplementedError
+
+    def build(self) -> Block:
+        if self._columns:
+            tables = [self._table_from_pydict(self._columns)]
+        else:
+            tables = []
+        tables.extend(self._tables)
+        if len(tables) > 1:
+            return self._concat_tables(tables)
+        elif len(tables) > 0:
+            return tables[0]
+        else:
+            return self._empty_table()
+
+    def num_rows(self) -> int:
+        return self._num_rows
+
+    def get_estimated_memory_usage(self) -> int:
+        if self._num_rows == 0:
+            return 0
+        return self._tables_size_bytes + self._uncompacted_size.size_bytes()
+
+    def _compact_if_needed(self) -> None:
+        assert self._columns
+        if self._uncompacted_size.size_bytes() < MAX_UNCOMPACTED_SIZE_BYTES:
+            return
+        block = self._table_from_pydict(self._columns)
+        self.add_block(block)
+        self._uncompacted_size = SizeEstimator()
+        self._columns.clear()
+        self._num_compactions += 1
+
+
+class TableBlockAccessor(BlockAccessor):
+    def __init__(self, table: Any):
+        self._table = table
+
+    def _create_table_row(self, row: Any) -> TableRow:
+        raise NotImplementedError
+
+    def iter_rows(self) -> Iterator[TableRow]:
+        outer = self
+
+        class Iter:
+            def __init__(self):
+                self._cur = -1
+
+            def __iter__(self):
+                return self
+
+            def __next__(self):
+                self._cur += 1
+                if self._cur < outer.num_rows():
+                    row = outer._create_table_row(
+                        outer.slice(self._cur, self._cur + 1, copy=False))
+                    return row
+                raise StopIteration
+
+        return Iter()
+
+    def _zip(self, acc: BlockAccessor) -> "Block[T]":
+        raise NotImplementedError
+
+    def zip(self, other: "Block[T]") -> "Block[T]":
+        acc = BlockAccessor.for_block(other)
+        if not isinstance(acc, type(self)):
+            raise ValueError("Cannot zip {} with block of type {}".format(
+                type(self), type(other)))
+        if acc.num_rows() != self.num_rows():
+            raise ValueError(
+                "Cannot zip self (length {}) with block of length {}".format(
+                    self.num_rows(), acc.num_rows()))
+        return self._zip(acc)
+
+    @staticmethod
+    def _empty_table() -> Any:
+        raise NotImplementedError
+
+    def _sample(self, n_samples: int, key: SortKeyT) -> Any:
+        raise NotImplementedError
+
+    def sample(self, n_samples: int, key: SortKeyT) -> Any:
+        if key is None or callable(key):
+            raise NotImplementedError(
+                "Table sort key must be a column name, was: {}".format(key))
```

Comment:
Use fstring instead.

Reply:
Can do. This line of code was moved from arrow_block.py.